routes/user.js

```diff
@@ -53,64 +53,111 @@ router.post(
+    check('password', 'Password required').isLength({ min: 6 }),
+    check('primaryLanguage', 'Please choose a language').isLength({ min: 1 }),
   ],
   runAsyncWrapper(async (req, res) => {
     const errors = validationResult(req);
     if (!errors.isEmpty())
       return res.status(422).json({ errors: errors.array() });
 
-    // Email must be unique
-    const user1 = await User.findOne({ email: req.body.email });
-    if (user1) {
-      return res.status(400).send('User with this email already exists.');
+    const { name, email, password, primaryLanguage } = req.body;
+
+    try {
+      // Email must be unique
+      let user = await User.findOne({ email });
+      //if we find a user they already have an account
+      if (user) {
+        return res.status(400).json({ msg: 'This user already exists' });
+      }
+
+      const salt = bcrypt.genSaltSync();
+      const hashed_password = bcrypt.hashSync(password, salt);
+
+      // REGISTER USER!!!
+      user = new User({
+        name,
+        email,
+        password: hashed_password,
+        primaryLanguage,
+      });
+
+      await user.save();
+
+      //Set up the jwt payload to user ID and email option
+      const payload = {
+        user: {
+          id: user._id,
+          email,
+        },
+      };
+
+      // success -> Get a JWT Token
+      //the payload is the data being encrypted
+      jwt.sign(
+        payload,
+        process.env.ACCESS_TOKEN_SECRET,
+        { expiresIn: JWT_EXPIRY_TIME },
+        (err, token) => {
+          if (err) throw err;
+          return (
+            res
+              .status(201)
+              // .cookie(token, { httpOnly: true })
+              .cookie('token', token)
+              .json({ token, msg: 'Register Success!' })
+          );
+        },
+      );
+    } catch (err) {
+      console.error(err.message);
+      res.status(500).send('Server Error');
     }
 
-    const salt = bcrypt.genSaltSync();
-    const hashed_password = bcrypt.hashSync(req.body.password, salt);
-    // REGISTER USER!!!
-    const user = await User.create({
-      name: req.body.name,
-      email: req.body.email,
-      password: hashed_password,
-      primaryLanguage: req.body.primaryLanguage,
-    });
-    // success -> Get a JWT Token
-    const email = req.body.email;
-    const accessToken = jwt.sign(
-      /* payload */ { email },
-      process.env.ACCESS_TOKEN_SECRET,
-      {
-        algorithm: 'HS256',
-        expiresIn: JWT_EXPIRY_TIME,
-      },
-    );
-    res.cookie('token', accessToken, { httpOnly: true });
-    //req.session.user = user
-    return res.status(201).send(user._id);
+   
   }),
 );
 
 router.get(
   '/get_by_id/:id',
   runAsyncWrapper(async function (req, res) {
     const user = await User.findById(req.params.id);
-    return res.status(200).json(user);
+
+    if (!user) {
+      return res.status(400).json({ error: 'User not found' });
+    } else {
+      return res.status(200).json(user);
```

Comment:
We probably don't want to include the user's password in the return here

Reply:
This is addressed at the root in the Users Model.   Bonnie added a .toJson and .toObject transformation method that deletes the password when a user is being retrieved

Reply:
Oh, I didn't realize that was called automatically with .json(user). :+1: